### PR TITLE
Skipping pytree registration in case fsdp is enabled

### LINKED ISSUE #39795: Regression - High memory usage when using transformers model with FSDP + LoRA
### System Info

- `transformers` version: 4.54.0
- Platform: Linux-5.14.0-284.73.1.el9_2.x86_64-x86_64-with-glibc2.35
- Python version: 3.10.13
- Huggingface_hub version: 0.34.1
- Safetensors version: 0.5.3
- Accelerate version: 1.9.0
- Accelerate config:    not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.7.1+cu126 (CUDA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: Yes, FSDP with accelerate
- Using GPU in script?: Yes
- GPU type: NVIDIA A100-SXM4-80GB

### Who can help?

@zach-huggingface @SunMarc

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

`sft.py`

```python
import torch
from accelerate import Accelerator
from transformers import AutoTokenizer, AutoModelForCausalLM
from peft import LoraConfig, get_peft_model
from peft.utils.other import fsdp_auto_wrap_policy


def main():

    model_name = "ibm-granite/granite-8b-code-base"

    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForCausalLM.from_pretrained(
        model_name,
        torch_dtype=torch.bfloat16,
    )
    dummy_input = tokenizer("This is a test sentence.", return_tensors="pt")

    accelerator = Accelerator()

    # if accelerator.is_main_process:
    #     torch.cuda.memory._record_memory_history(max_entries=100000)

    peft_config = LoraConfig(
        r=4,
        lora_alpha=16,
        lora_dropout=0.05,
        bias="none",
        target_modules=["q_proj", "v_proj"]
    )

    model = get_peft_model(model, peft_config)

    fsdp_plugin = accelerator.state.fsdp_plugin
    fsdp_plugin.auto_wrap_policy = fsdp_auto_wrap_policy(model)  # type: ignore

    if accelerator.is_main_process:
        model.print_trainable_parameters()

    model = accelerator.prepare(model)
    optimizer = torch.optim.AdamW(model.parameters(), lr=5e-5)
    optimizer = accelerator.prepare(optimizer)

    model.train()

    torch.cuda.empty_cache()

    accelerator.print(f"Memory allocated after setup: {torch.cuda.memory_allocated() / 1e9:.2f} GB")

    outputs = model(**dummy_input, labels=dummy_input["input_ids"])

    loss = outputs.loss
    accelerator.backward(loss)
    optimizer.step()
    optimizer.zero_grad()

    peak_memory = torch.cuda.max_memory_allocated() / 1e9

    accelerator.print(f"Peak memory during training step: {peak_memory:.2f} GB")

    accelerator.wait_for_everyone()
    accelerator.print("Debug script finished successfully.")

    # if accelerator.is_main_process:
    #     torch.cuda.memory._dump_snapshot("profile_449.pkl")
    #     torch.cuda.memory._record_memory_history(enabled=None)

if __name__ == "__main__":
    """
    accelerate launch --config_file fsdp.yaml -m sft
    """
    main()
```

`fsdp.yaml`
```
compute_environment: LOCAL_MACHINE
distributed_type: FSDP
fsdp_config:
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
  fsdp_backward_prefetch: BACKWARD_PRE
  fsdp_forward_prefetch: false
  fsdp_offload_params: false
  fsdp_sharding_strategy: FULL_SHARD
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_cpu_ram_efficient_loading: false
  fsdp_sync_module_states: true
  fsdp_use_orig_params: true
mixed_precision: 'no'
machine_rank: 0
num_machines: 1
num_processes: 4
rdzv_backend: static
same_network: true
```

Run

```
accelerate launch --config_file fsdp.yaml -m sft
```

### Expected behavior

When I am doing LoRA fine-tuning with FSDP, I am seeing a huge memory usage compared to transformers v4.49.0. This issue is specific to versions including 4.50.0 and above. For example, 

For 4 GPUs, I see the following memory usage on `transformers==4.49.0`
```
Memory allocated after setup: 4.03 GB
Peak memory during training step: 5.36 GB
```

vs when I am using any higher version `transformers==4.54.0`
```
Memory allocated after setup: 4.03 GB
Peak memory during training step: 20.16 GB
```

The peak memory usage is 4x.

Keeping all other library versions constant, the bug only appears when upgrading transformers to any version above 4.49.0. That's the reason I have raised the bug here and not in accelerate. Downgrading to `transformers==4.49.0` fixes the issue. 

The issue ends here, but I will provide some of my findings in case it is helpful
1. I was able to reproduce this issue in other Llama-based models, too.
2. The bug only appears with FSDP + LoRA. Single GPU jobs don't seem to have the bug.
3. I have already tried the solution provided here: https://github.com/huggingface/accelerate/issues/3474 and it does not solve the issue
4. The memory explosion happens during the backward pass, specifically at: `accelerator.backward(loss)`
5. Looking at the memory profiling results, it _seems_ like all attention heads (Q, V) are somehow treated as trainable and the memory is reserved for their optimizer states which is leading to this 4x spike. I am also attaching the photos from the memory profiling.
6. For fsdp config, I have tried both values of - `fsdp_cpu_ram_efficient_loading`, `fsdp_use_orig_params`, with and without setting `fsdp_transformer_layer_cls_to_wrap`

Image
Image

### PR DESCRIPTION
**Pull request by romitjain**
# What does this PR do?

Fixes #39795 

Pytree registration for Dynamic KV Cache prevents FSDP from sharding the Q, K, V cache matrices correctly. This leads to a memory explosion during training (see the attached issue).

This PR fixes it by moving the pytree registration of Dynamic KV Cache from `cache_utils` to `integrations.executorch`, where it is only executed in case `torch.export` is called. This is an improvement over the previous version, where this registration was happening globally.

As a side effect of this PR, I have also moved `is_fsdp_enabled` helper function to a more suitable location -> `integrations.fsdp` (from `modeling_utils`)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #39795 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@Cyrilvallez @SunMarc @gante

### CODE REVIEW

**Comment by Cyrilvallez** [comment]
cc @SunMarc  @gante as well, looks fine to me as the pytree registration was added for ddp/export support, but just want to make sure

@romitjain you will need to move `is_fsdp_enabled` outside of `modeling_utils.py` to avoid circular imports, e.g. `utils/imports_utils.py` would be a good place - but are you sure the env variables inside `is_fsdp_enabled` will be correctly set when importing `cache_utils`? cc @SunMarc on this as well

**Review by gante — APPROVED** [review]
LGTM (after fixing import issues, see above), thank you for fixing 🤗 

related: [my reply in the original issue](https://github.com/huggingface/transformers/issues/39795#issuecomment-3179793156), this is indeed the way to go about the original issue

**Comment by romitjain** [comment]
@gante @Cyrilvallez 

1. I have moved `is_fsdp_enabled` to integrations.fsdp to avoid circular import
2. I have moved PyTree registration inside the `DynamicCache` class since it was getting registered globally at the imports earlier. The first indication that the model is being run in FSDP is during its forward pass. During that time, it also creates the `DynamicCache` class (https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L371). Hence, this is the appropriate location to decide if we want to register pytees. It's controlled by class variables and methods, so it won't lead to duplicate registrations.

PS: My validations are passing now. The memory usage is back to what is expected and stated in the original issue.

Please have a look.

**Comment by gante** [comment]
@romitjain this is indeed a valid alternative, but let me double-check the best design with @Cyrilvallez 🤗 If we move on with this design, we would need to apply the pattern to the hybrid cache as well.

_______________________________

@Cyrilvallez 

TL;DR -- The root problem here is that we register the caches at import time for `torch.export`, and we can only detect FSDP (which can't have the registered caches) at forward time.

I see two solutions:
1. [latest commit by @romitjain] we register the caches at cache instance initialization time. This should work because our executorch integration initializes the needed cache instance before exporting the model.
2. we have a function to register the caches, which needs to be manually called by folks before exporting.

Both are BC breaking in the sense that users with custom executorch usage will likely need to change their code anyway. 2. is slightly more breaking, as it also requires changes in our own scripts, while I expect 1. to work without changes. However, 2. is much more explicit to control, which I think will prevent issues down the line -- for instance, if we go with 1. and we manually initialize a cache before using anything that allows us to detect FSDP, we get the original issue again. Therefore, I would prefer to go with 2.

WDYT?

**Comment by Cyrilvallez** [comment]
Yup, totally agree @gante, I don't like having it inside the Cache either, issue would probably come back at some point. However, maybe a function is not strictly needed? If you can confirm that it's used ONLY for export, then let's register it either at the top of `integrations.executorch`, or if still too restrictive, inside [this function](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/executorch.py#L975-L993)? Then it would only be run if export is actually used!

Because due to `vmap` limitations, we need to export with a specific function anyway even for DynamicCache!

That way, it's actually entirely part of `export` instead of living in `cache_utils` as well, which I like a lot actually, as stumbling upon this hard-to-read code in the middle of the Cache is not ideal

**Comment by romitjain** [comment]
@Cyrilvallez @gante 
Thanks for the feedback. Sounds good to me to register it specifically under the function you pointed out. If it's good to go ahead from your side, let me know, and I can make the changes and push the commit.

**Comment by gante** [comment]
@romitjain: @Cyrilvallez and I agree with the plan so, if you're up for it, we'll greatly appreciate your commits with the changes 💛 

As Cyril suggested, if it is compatible with the FSDP case, then registering at the top of `integrations.executorch` would be ideal.

**Comment by romitjain** [comment]
@gante This is ready for review now.

Essentially, I have moved the pytree registration to `integrations.executorch`, which will be called only when exporting with `DynamicCache`.

I avoided the global registration since `integrations.executorch` functions are being called in `src.transformers.__init__`

**Review by gante — APPROVED** [review]
Yup, that's it -- thank you for iterating 💛

**Comment by gante** [comment]
@Cyrilvallez have a look at the PR too please 🤗

**Review by Cyrilvallez — APPROVED** [review]
LGTM! Thanks for the great cleanup @romitjain 🤗 Beyond fixing fsdp, it makes the pytree registration much clearer 🤗

**Review by SunMarc — COMMENTED** [review]
Nice job !
